godot-cell/src/cell.rs

```diff
@@ -0,0 +1,353 @@
+        let mut state = self.state.lock().unwrap();
+        state.borrow_state.increment_mut()?;
+        let count = state.borrow_state.mut_count();
+        let value = state.get_ptr();
+
+        // SAFETY: `increment_mut` succeeded, therefore any existing mutable references are inaccessible.
+        // Additionally no new references can be created, unless the returned guard is made inaccessible.
+        //
+        // This is the case because the only way for a new `GdMut` or `GdRef` to be made after this is for
+        // either this guard to be dropped or `make_inaccessible` to be called and succeed.
+        //
+        // If this guard is dropped, then we don't need to worry.
+        //
+        // If `make_inaccessible` is called and succeeds, then a mutable reference from this guard is passed
+        // in. In which case, we cannot use this guard again until the resulting inaccessible guard is
+        // dropped.
+        unsafe { Ok(MutGuard::new(&self.get_ref().state, count, value)) }
+    }
+
+    /// Make the current mutable borrow inaccessible, thus freeing the value up to be reborrowed again.
+    ///
+    /// Will error if:
+    /// - There is currently no accessible mutable borrow.
+    /// - There are any shared references.
+    /// - `current_ref` is not equal to the pointer in `self.state`.
+    pub fn make_inaccessible<'cell: 'val, 'val>(
+        self: Pin<&'cell Self>,
+        current_ref: &'val mut T,
+    ) -> Result<InaccessibleGuard<'val, T>, Box<dyn Error>> {
+        InaccessibleGuard::new(&self.get_ref().state, current_ref)
+    }
+
+    /// Returns `true` if there are any mutable or shared references, regardless of whether the mutable
+    /// references are accessible or not.
+    ///
+    /// In particular this means that it is safe to destroy this cell and the value contained within, as no
+    /// references can exist that can reference this cell.
+    ///
+    /// Keep in mind that in multithreaded code it is still possible for this to return true, and then the
+    /// cell hands out a new borrow before it is destroyed. So we still need to ensure that this cannot
+    /// happen at the same time.
+    pub fn is_currently_bound(self: Pin<&Self>) -> bool {
+        let state = self.state.lock().unwrap();
+
+        state.borrow_state.shared_count() > 0 || state.borrow_state.mut_count() > 0
+    }
+
+    /// Similar to [`Self::is_currently_bound`] but only counts mutable references and ignores shared references.
+    pub(crate) fn is_currently_mutably_bound(self: Pin<&Self>) -> bool {
+        let state = self.state.lock().unwrap();
+
+        state.borrow_state.mut_count() > 0
+    }
+}
+
+// SAFETY: `T` is sync so we can return references to it on different threads.
+// Additionally all internal state is synchronized via a mutex, so we wont have race conditions when trying
+// to use it from multiple threads.
+unsafe impl<T: Sync> Sync for GdCellInner<T> {}
```

Comment:
```suggestion
// SAFETY: `T` is sync so we can return references to it on different threads, it is also send so we can return
// mutable references to it on different threads.
// Additionally all internal state is synchronized via a mutex, so we wont have race conditions when trying
// to use it from multiple threads.
unsafe impl<T: Send + Sync> Sync for GdCellInner<T> {}
```

Just now noticed that we need a `Send` bound here as well. Since we can use `borrow_mut()` to get a mutable reference on a different thread then `std::mem::swap` to move the value to the new thread.